### PR TITLE
Add test bash scripts for compiler tests and integrate with Travis

### PR DESCRIPTION
**Pull request by KentShikama**
Addresses https://rchain.atlassian.net/browse/RHOL-21

Note this breaks the build for Mac (but according to @stay, it is already broken anyways). I think we should all use vagrant with Ubuntu 14.04 to get the same environment for the C++ build.

### CODE REVIEW

**Comment by KentShikama** [comment]
Build passes :) (https://travis-ci.org/rchain/rchain/builds/300063813). We should integrate the PR system with Travis just like with "Review required".

**Comment by pyrocto** [comment]
It's not feasible to build only on Ubuntu.  Greg uses Macs exclusively, and is going to want to demo this thing from his own laptop and desktop.

File: rholang/test.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+sbt bnfc:generate
+sbt assembly
+# Run compiler over all test rholang files
+for i in tests/*; do
+	./rho2rbl "$i"
+done
+# Rewrite all tests to print to a text file results.txt instead of stdout
```
Comment (pyrocto) [inline rholang/test.sh]: This is an impressive kludge.

File: rosette/rbl/rosette/tests/arithmetic_test.rbl [hunk 2]
```diff
@@ -0,0 +1 @@
+(if (match? [(+ 2 8) (+ (+ (+ (+ (+ (+ (+ (+ (+ 1 2) 3) 4) 5) 6) 7) 8) 9) 10) (+ 2147483647 2147483647) (+ 2.5 7.5) (+ 3.141592653589793 3.141592653589793) (- 47 37) (- (- 2147483647) 1) (* 2 5) (* (- 5) (- 2)) (* 2147483647 2147483647) (/ 120 12) (/ 2147483647 (- 1)) (+ (- (- (+ (- 10) (* 4 5)) (/ 30 10)) 2) 5) (+ 10 (* (- 10 10) 10)) (- (+ (- 10.5 0.5) (/ 2.5 2.5)) 1.0)] [10 55 -2 10.0 6.283185307179586 10 0 10 10 1 10 1 10 10 10.0]) (seq (ostream-print (ostream-new "results.txt") "Pass") (display #\\n)) (seq (ostream-print (ostream-new "results.txt") "Fail") (display #\\n)))
```
Comment (pyrocto) [inline rosette/rbl/rosette/tests/arithmetic_test.rbl]: Any reason not to format this nicely?

File: rosette/h/Addr.h [hunk 3]
```diff
@@ -38,8 +38,8 @@ int  addr_to_pre_fixnum(Ob *);
 #define ADDR_TO_PRE_FIXNUM(x)  nontrivial_addr_to_pre_fixnum(x)
 #define PRE_FIXNUM_TO_ADDR(x) nontrivial_pre_fixnum_to_addr(x)
 #else
-#define ADDR_TO_PRE_FIXNUM(x) ((Word32)(intptr_t)(x))
-#define PRE_FIXNUM_TO_ADDR(x) ((intptr_t)(Ob *)(x))
+#define ADDR_TO_PRE_FIXNUM(x) ((Word32)(int)(x))
```
Comment (pyrocto) [inline rosette/h/Addr.h]: Can you explain why intptr_t needs to be replaced by int?

**Comment by KentShikama** [comment]
I understand; I'll add a ticket to build on a Mac. If I understand correctly, it doesn't build on Mac as of now (before this PR) anyways. 

On November 10, 2017 2:03:25 PM PST, Mike Stay <notifications@github.com> wrote:
>It's not feasible to build only on Ubuntu.  Greg uses Macs exclusively,
>and is going to want to demo this thing from his own laptop and
>desktop.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/rchain/rchain/pull/52#issuecomment-343599299

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by KentShikama** [comment]
I'm not sure. This is what it was originally (30 years ago). I would like to ask Henry why an int should be turned into an intptr_t?

On November 10, 2017 2:08:53 PM PST, Mike Stay <notifications@github.com> wrote:
>pyrocto commented on this pull request.
>
>
>
>> @@ -0,0 +1,10 @@
>+sbt bnfc:generate
>+sbt assembly
>+# Run compiler over all test rholang files
>+for i in tests/*; do
>+	./rho2rbl "$i"
>+done
>+# Rewrite all tests to print to a text file results.txt instead of
>stdout
>
>This is an impressive kludge.
>
>> @@ -0,0 +1 @@
>+(if (match? [(+ 2 8) (+ (+ (+ (+ (+ (+ (+ (+ (+ 1 2) 3) 4) 5) 6) 7) 8)
>9) 10) (+ 2147483647 2147483647) (+ 2.5 7.5) (+ 3.141592653589793
>3.141592653589793) (- 47 37) (- (- 2147483647) 1) (* 2 5) (* (- 5) (-
>2)) (* 2147483647 2147483647) (/ 120 12) (/ 2147483647 (- 1)) (+ (- (-
>(+ (- 10) (* 4 5)) (/ 30 10)) 2) 5) (+ 10 (* (- 10 10) 10)) (- (+ (-
>10.5 0.5) (/ 2.5 2.5)) 1.0)] [10 55 -2 10.0 6.283185307179586 10 0 10
>10 1 10 1 10 10 10.0]) (seq (ostream-print (ostream-new "results.txt")
>"Pass") (display #\\n)) (seq (ostream-print (ostream-new "results.txt")
>"Fail") (display #\\n)))
>
>Any reason not to format this nicely?
>
>> @@ -38,8 +38,8 @@ int  addr_to_pre_fixnum(Ob *);
> #define ADDR_TO_PRE_FIXNUM(x)  nontrivial_addr_to_pre_fixnum(x)
> #define PRE_FIXNUM_TO_ADDR(x) nontrivial_pre_fixnum_to_addr(x)
> #else
>-#define ADDR_TO_PRE_FIXNUM(x) ((Word32)(intptr_t)(x))
>-#define PRE_FIXNUM_TO_ADDR(x) ((intptr_t)(Ob *)(x))
>+#define ADDR_TO_PRE_FIXNUM(x) ((Word32)(int)(x))
>
>Can you explain why intptr_t needs to be replaced by int?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/rchain/rchain/pull/52#pullrequestreview-75893704

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: rholang/test.sh (hunk 1 above)
Comment (KentShikama) [reply rholang/test.sh]: I thought this was the simplest way to handle things. Any better ideas?

File: rosette/rbl/rosette/tests/arithmetic_test.rbl (hunk 2 above)
Comment (KentShikama) [reply rosette/rbl/rosette/tests/arithmetic_test.rbl]: Whoops, thanks for the catch. That shouldn't be checked in.

**Comment by pyrocto** [comment]
> If I understand correctly, it doesn't build on Mac as of now (before this PR) anyways.

This branch builds fine on Mac:
https://github.com/rchain/Rosette/tree/fix-build
I haven't done a recursive diff to find out what's different in the version from this PR.

File: rholang/test.sh (hunk 1 above)
Comment (pyrocto) [reply rholang/test.sh]: It's the simplicity that's impressive! :D The "right" way would be to have the VM able to read commandline flags and pass them into contracts to use; then you'd have the compiler output `ostream-print (ostream-new args[0])` directly.  But this is fine for the moment.

**Comment by KentShikama** [comment]
Yeah ok, I guess that the Mac build needs to be fixed in this PR then. Long shot but would you be open to having everyone just use vagrant (or docker)? Because technically this doesn't even build on Ubuntu 16.04, only 14.04. And so I feel like even if we get it building on Mac and Ubuntu 14.04, there will be other OSs or versions of OSs that we haven't accounted for.

**Comment by pyrocto** [comment]
I'd be fine with just one version of Ubuntu, except that Greg uses OS X.  So make it work on OS X and somewhere else, and we can have docker or whatever for that one other place.

**Comment by dckc** [comment]
This also addresses ROS-271, as far as I can tell.

The biggest problem I can see is that the [travis logs](https://travis-ci.org/rchain/rchain) are hard to read: "This log is too long to be displayed. Please reduce the verbosity of your build ..." But if I burrow into the raw log, I do see:

```
The command "cd rholang && bash test.sh && cd ../rosette && bash test.sh" exited with 0.
```

**Comment by dckc** [comment]
As discussed with @KentShikama , I plan to add the "The build runs a build script, that will try to figure out which directories have changed" script.

Medha is following up separately on the Mac build issue.

**Comment by dckc** [comment]
I made some progress in #54 but I'm blocked by #58.

#55, #57 don't help either.

**Comment by DoSOfRedRiver** [comment]
I think we should change this according to #54 
Also, you may want to add some settings to .travis.yml like caching your dependencies etc.

**Comment by dckc** [comment]
@DoSOfRedRiver if you have pointers on how to cache dependencies, I'd appreciate it.

**Comment by DoSOfRedRiver** [comment]
Sure! 
For example, for Scala you can use something like:
```
cache:
  directories:
  - "$HOME/.ivy2/cache"
  - "$HOME/.sbt"
```
Please, also take a look at https://docs.travis-ci.com/user/caching/

**Comment by dckc** [comment]
I did the caching in 4743ad5; see #60.

**Comment by MParlikar** [comment]
We should not leave PR's open, changes have been requested, and now we have conflicts.  Fix this PR or close it.

**Comment by dckc** [comment]
Now that #60 is merged, this should be closed.

**Comment by KentShikama** [comment]
@dckc Thanks for completing!

**Pull request closed without merging**